diff --git a/src/observability/reports/SeverityLevel.py b/src/observability/reports/SeverityLevel.py
index b61f3ac..d58e1ea 100644
--- a/src/observability/reports/SeverityLevel.py
+++ b/src/observability/reports/SeverityLevel.py
@@ -26,10 +26,15 @@ from enum import Enum
 
 __version__ = VERSION
 
+
+
+
+
+
 class SeverityLevel(Enum):
     """Severity level for issues."""
 
     INFO = 1
     WARNING = 2
     ERROR = 3
-    CRITICAL = 4
\ No newline at end of file
+    CRITICAL = 4
