diff --git a/src/infrastructure/orchestration/SignalBusOrchestrator.py b/src/infrastructure/orchestration/SignalBusOrchestrator.py
index 1fcf490..710355b 100644
--- a/src/infrastructure/orchestration/SignalBusOrchestrator.py
+++ b/src/infrastructure/orchestration/SignalBusOrchestrator.py
@@ -23,7 +23,7 @@ from src.core.base.version import VERSION
 import logging
 import queue
 import threading
-from typing import Dict, List, Any
+from typing import Any
 from collections.abc import Callable
 
 __version__ = VERSION
@@ -72,4 +72,4 @@ class SignalBusOrchestrator:
         """Stops the signal bus."""
         self._running = False
         self._thread.join()
-        logging.info("SignalBus: Shutdown complete.")
\ No newline at end of file
+        logging.info("SignalBus: Shutdown complete.")
