diff --git a/src/logic/agents/intelligence/DataAgent.py b/src/logic/agents/intelligence/DataAgent.py
index 45723e7..4bca6fb 100644
--- a/src/logic/agents/intelligence/DataAgent.py
+++ b/src/logic/agents/intelligence/DataAgent.py
@@ -118,4 +118,4 @@ class DataAgent(BaseAgent):
 
 if __name__ == "__main__":
     main = create_main_function(DataAgent, "Data Agent", "Path to data log")
-    main()
\ No newline at end of file
+    main()
