diff --git a/src/logic/agents/development/SecurityGuardAgent.py b/src/logic/agents/development/SecurityGuardAgent.py
index 7aef943..40e50d0 100644
--- a/src/logic/agents/development/SecurityGuardAgent.py
+++ b/src/logic/agents/development/SecurityGuardAgent.py
@@ -25,7 +25,6 @@ from src.core.base.version import VERSION
 from src.core.base.BaseAgent import BaseAgent
 from src.core.base.utilities import create_main_function
 from src.logic.agents.development.SecurityCore import SecurityCore
-from typing import List, Tuple
 
 __version__ = VERSION
 
