diff --git a/src/logic/tools/evolved/sampletask.py b/src/logic/tools/evolved/sampletask.py
index a815fda..a63374c 100644
--- a/src/logic/tools/evolved/sampletask.py
+++ b/src/logic/tools/evolved/sampletask.py
@@ -32,4 +32,4 @@ def sample_automated_task() -> None:
     pyautogui.click(100, 200)
     pyautogui.press('a')
     pyautogui.press('enter')
-    pyautogui.click(150, 250)
\ No newline at end of file
+    pyautogui.click(150, 250)
