diff --git a/src/core/base/utils/NotificationCore.py b/src/core/base/utils/NotificationCore.py
index 4a29567..4aef7d2 100644
--- a/src/core/base/utils/NotificationCore.py
+++ b/src/core/base/utils/NotificationCore.py
@@ -28,7 +28,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 import time
 import urllib.parse
-from typing import Dict, Any
+from typing import Any
 
 __version__ = VERSION
 
