diff --git a/src/infrastructure/dev/test_utils/FixtureGenerator.py b/src/infrastructure/dev/test_utils/FixtureGenerator.py
index 16236d1..2dbc6ce 100644
--- a/src/infrastructure/dev/test_utils/FixtureGenerator.py
+++ b/src/infrastructure/dev/test_utils/FixtureGenerator.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .TestFixture import TestFixture
 from pathlib import Path
-from typing import Dict, Optional
 import logging
 import shutil
 import tempfile
