diff --git a/src/logic/agents/cognitive/TheoryOfMind.py b/src/logic/agents/cognitive/TheoryOfMind.py
index 7eaf818..42b62e6 100644
--- a/src/logic/agents/cognitive/TheoryOfMind.py
+++ b/src/logic/agents/cognitive/TheoryOfMind.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from typing import Dict, Any, List
-from src.logic.cognitive.TheoryOfMindCore import TheoryOfMindCore
+from src.logic.agents.cognitive.TheoryOfMindCore import TheoryOfMindCore
 
 __version__ = VERSION
 
