diff --git a/src/infrastructure/orchestration/RoutingEngine.py b/src/infrastructure/orchestration/RoutingEngine.py
index 96cc057..dddc84a 100644
--- a/src/infrastructure/orchestration/RoutingEngine.py
+++ b/src/infrastructure/orchestration/RoutingEngine.py
@@ -1,7 +1,7 @@
 
 import logging
 import os
-from typing import Any, Dict
+from typing import Any
 from src.infrastructure.backend.RunnerBackends import BackendHandlers
 
 class RoutingEngine:
