diff --git a/src/infrastructure/dev/test_utils/PerformanceTracker.py b/src/infrastructure/dev/test_utils/PerformanceTracker.py
index ef4a9e2..9b091d5 100644
--- a/src/infrastructure/dev/test_utils/PerformanceTracker.py
+++ b/src/infrastructure/dev/test_utils/PerformanceTracker.py
@@ -25,7 +25,7 @@ from src.core.base.version import VERSION
 from .PerformanceMetric import PerformanceMetric
 from .PerformanceMetricType import PerformanceMetricType
 from contextlib import contextmanager
-from typing import Any, Dict, List
+from typing import Any
 from collections.abc import Iterator
 import time
 
