diff --git a/src/logic/agents/cognitive/StrategicPlanningAgent.py b/src/logic/agents/cognitive/StrategicPlanningAgent.py
index ed4c2ed..67ecb10 100644
--- a/src/logic/agents/cognitive/StrategicPlanningAgent.py
+++ b/src/logic/agents/cognitive/StrategicPlanningAgent.py
@@ -19,7 +19,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict, List, Any
+from typing import Any
 from src.core.base.BaseAgent import BaseAgent
 
 __version__ = VERSION
