diff --git a/src/infrastructure/fleet/SecretManager.py b/src/infrastructure/fleet/SecretManager.py
index 58245a6..d2d1dd1 100644
--- a/src/infrastructure/fleet/SecretManager.py
+++ b/src/infrastructure/fleet/SecretManager.py
@@ -27,7 +27,6 @@ from src.core.base.version import VERSION
 import os
 import json
 import logging
-from typing import Optional
 from .SecretCore import SecretCore
 
 __version__ = VERSION
