diff --git a/src/infrastructure/dev/test_utils/AgentAssertions.py b/src/infrastructure/dev/test_utils/AgentAssertions.py
index a36db34..ad80287 100644
--- a/src/infrastructure/dev/test_utils/AgentAssertions.py
+++ b/src/infrastructure/dev/test_utils/AgentAssertions.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .TestAssertion import TestAssertion
-from typing import List
 import json
 import re
 
