diff --git a/src/infrastructure/backend/QueuedRequest.py b/src/infrastructure/backend/QueuedRequest.py
index 166e868..6e09b8d 100644
--- a/src/infrastructure/backend/QueuedRequest.py
+++ b/src/infrastructure/backend/QueuedRequest.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass
-from typing import Optional
 from collections.abc import Callable
 
 __version__ = VERSION
