diff --git a/src/logic/agents/development/PerformanceAgent.py b/src/logic/agents/development/PerformanceAgent.py
index d855ea7..9a7548a 100644
--- a/src/logic/agents/development/PerformanceAgent.py
+++ b/src/logic/agents/development/PerformanceAgent.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.types.OptimizationSuggestion import OptimizationSuggestion
 from src.core.base.types.OptimizationType import OptimizationType
-from typing import List, Tuple
 import re
 
 __version__ = VERSION
