diff --git a/src/logic/agents/cognitive/context/utils/ContextRecommender.py b/src/logic/agents/cognitive/context/utils/ContextRecommender.py
index 64b07f5..791af93 100644
--- a/src/logic/agents/cognitive/context/utils/ContextRecommender.py
+++ b/src/logic/agents/cognitive/context/utils/ContextRecommender.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from src.logic.agents.cognitive.context.models.ContextRecommendation import ContextRecommendation
-from typing import Dict, List, Optional, Tuple
 import re
 
 __version__ = VERSION
