diff --git a/src/infrastructure/backend/SystemResponse.py b/src/infrastructure/backend/SystemResponse.py
index 5853d8b..ad5efb6 100644
--- a/src/infrastructure/backend/SystemResponse.py
+++ b/src/infrastructure/backend/SystemResponse.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass
-from typing import Optional
 
 __version__ = VERSION
 
