diff --git a/src/core/modules/CodeQualityModule.py b/src/core/modules/CodeQualityModule.py
index 2111905..56ee6b3 100644
--- a/src/core/modules/CodeQualityModule.py
+++ b/src/core/modules/CodeQualityModule.py
@@ -13,7 +13,7 @@
 
 from __future__ import annotations
 import re
-from typing import Any, Dict, List
+from typing import Any
 from src.core.base.modules import BaseModule
 
 class CodeQualityModule(BaseModule):
