diff --git a/src/observability/reports/ChangelogTemplate.py b/src/observability/reports/ChangelogTemplate.py
index fe26449..abd566a 100644
--- a/src/observability/reports/ChangelogTemplate.py
+++ b/src/observability/reports/ChangelogTemplate.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
