diff --git a/src/observability/stats/exporters/CloudExporter.py b/src/observability/stats/exporters/CloudExporter.py
index eb8a198..859701c 100644
--- a/src/observability/stats/exporters/CloudExporter.py
+++ b/src/observability/stats/exporters/CloudExporter.py
@@ -25,7 +25,7 @@ from src.core.base.version import VERSION
 from ..observability_core import ExportDestination
 from ..observability_core import Metric
 from datetime import datetime
-from typing import Any, Dict, List, Optional
+from typing import Any
 import json
 import logging
 
