diff --git a/src/logic/agents/development/ProfilingAgent.py b/src/logic/agents/development/ProfilingAgent.py
index 203fe65..82f0590 100644
--- a/src/logic/agents/development/ProfilingAgent.py
+++ b/src/logic/agents/development/ProfilingAgent.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.types.ProfilingCategory import ProfilingCategory
 from src.core.base.types.ProfilingSuggestion import ProfilingSuggestion
-from typing import Any, List
+from typing import Any
 import ast
 import logging
 import pstats
