diff --git a/src/logic/agents/cognitive/LogicProverAgent.py b/src/logic/agents/cognitive/LogicProverAgent.py
index a76f935..32bb9ec 100644
--- a/src/logic/agents/cognitive/LogicProverAgent.py
+++ b/src/logic/agents/cognitive/LogicProverAgent.py
@@ -19,7 +19,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict, List, Any
+from typing import Any
 
 __version__ = VERSION
 
