diff --git a/src/observability/reports/ReportAPI.py b/src/observability/reports/ReportAPI.py
index d529d63..1cbdcb3 100644
--- a/src/observability/reports/ReportAPI.py
+++ b/src/observability/reports/ReportAPI.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ReportType import ReportType
 from pathlib import Path
-from typing import List, Optional
 import logging
 
 __version__ = VERSION
