diff --git a/src/infrastructure/backend/RequestBatcher.py b/src/infrastructure/backend/RequestBatcher.py
index d38e948..ce85553 100644
--- a/src/infrastructure/backend/RequestBatcher.py
+++ b/src/infrastructure/backend/RequestBatcher.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .BatchRequest import BatchRequest
-from typing import List, Optional
 import threading
 import time
 from src.infrastructure.backend.LocalContextRecorder import LocalContextRecorder
