diff --git a/src/observability/reports/ReportSubscription.py b/src/observability/reports/ReportSubscription.py
index fd15147..53ad24c 100644
--- a/src/observability/reports/ReportSubscription.py
+++ b/src/observability/reports/ReportSubscription.py
@@ -25,7 +25,6 @@ from src.core.base.version import VERSION
 from .ReportType import ReportType
 from .SubscriptionFrequency import SubscriptionFrequency
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
