diff --git a/src/infrastructure/backend/TTLCache.py b/src/infrastructure/backend/TTLCache.py
index a4c66fa..b0aec4d 100644
--- a/src/infrastructure/backend/TTLCache.py
+++ b/src/infrastructure/backend/TTLCache.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .CachedResponse import CachedResponse
-from typing import Any, Dict, Optional
+from typing import Any
 import threading
 import time
 
