diff --git a/src/core/base/utils/TemplateManager.py b/src/core/base/utils/TemplateManager.py
index 4dce112..be2e5bf 100644
--- a/src/core/base/utils/TemplateManager.py
+++ b/src/core/base/utils/TemplateManager.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.utils.AgentTemplate import AgentTemplate
-from typing import List, Optional, Dict
 
 __version__ = VERSION
 
