diff --git a/src/infrastructure/dev/test_utils/EnvironmentDetector.py b/src/infrastructure/dev/test_utils/EnvironmentDetector.py
index 252e547..4c103aa 100644
--- a/src/infrastructure/dev/test_utils/EnvironmentDetector.py
+++ b/src/infrastructure/dev/test_utils/EnvironmentDetector.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Any, Dict
+from typing import Any
 
 __version__ = VERSION
 
