diff --git a/src/core/base/utils/ScheduledExecution.py b/src/core/base/utils/ScheduledExecution.py
index 63ad5b6..267b19e 100644
--- a/src/core/base/utils/ScheduledExecution.py
+++ b/src/core/base/utils/ScheduledExecution.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.utils._helpers import _empty_dict_str_any
 from dataclasses import dataclass, field
-from typing import Optional, Any
+from typing import Any
 
 __version__ = VERSION
 
