diff --git a/src/infrastructure/api/ImportedEntry.py b/src/infrastructure/api/ImportedEntry.py
index 867d1ee..c4adf64 100644
--- a/src/infrastructure/api/ImportedEntry.py
+++ b/src/infrastructure/api/ImportedEntry.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ImportSource import ImportSource
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
