diff --git a/src/core/base/types/OptimizationType.py b/src/core/base/types/OptimizationType.py
index de417b3..1097c3b 100644
--- a/src/core/base/types/OptimizationType.py
+++ b/src/core/base/types/OptimizationType.py
@@ -26,10 +26,20 @@ from enum import Enum
 
 __version__ = VERSION
 
+
+
+
+
+
+
+
+
+
+
 class OptimizationType(Enum):
     """Types of code optimization."""
     ALGORITHMIC = "algorithmic"
     MEMORY = "data/memory"
     IO = "io"
     CONCURRENCY = "concurrency"
-    CACHING = "caching"
\ No newline at end of file
+    CACHING = "caching"
