diff --git a/src/observability/improvements/ArchivedImprovement.py b/src/observability/improvements/ArchivedImprovement.py
index f4056af..1c433da 100644
--- a/src/observability/improvements/ArchivedImprovement.py
+++ b/src/observability/improvements/ArchivedImprovement.py
@@ -27,6 +27,11 @@ from dataclasses import dataclass
 
 __version__ = VERSION
 
+
+
+
+
+
 @dataclass
 class ArchivedImprovement:
     """An archived improvement.
@@ -40,4 +45,4 @@ class ArchivedImprovement:
     improvement: Improvement
     archived_date: str = ""
     archived_by: str = ""
-    archive_reason: str = ""
\ No newline at end of file
+    archive_reason: str = ""
