diff --git a/src/observability/errors/ErrorCategory.py b/src/observability/errors/ErrorCategory.py
index da7b766..34edc10 100644
--- a/src/observability/errors/ErrorCategory.py
+++ b/src/observability/errors/ErrorCategory.py
@@ -26,6 +26,11 @@ from enum import Enum
 
 __version__ = VERSION
 
+
+
+
+
+
 class ErrorCategory(Enum):
     """Error categories."""
     SYNTAX = "syntax"
@@ -38,4 +43,4 @@ class ErrorCategory(Enum):
     DEPRECATION = "deprecation"
     VALUE = "value"
     IMPORT = "import"
-    OTHER = "other"
\ No newline at end of file
+    OTHER = "other"
