diff --git a/src/infrastructure/dev/test_utils/TestDataType.py b/src/infrastructure/dev/test_utils/TestDataType.py
index c4a30d4..435591a 100644
--- a/src/infrastructure/dev/test_utils/TestDataType.py
+++ b/src/infrastructure/dev/test_utils/TestDataType.py
@@ -26,12 +26,17 @@ from enum import Enum
 
 __version__ = VERSION
 
+
+
+
+
+
 class TestDataType(Enum):
-    __test__ = False
     """Types of test data."""
+    __test__ = False
 
     PYTHON_CODE = "python_code"
     MARKDOWN = "markdown"
     JSON = "json"
     YAML = "yaml"
-    TEXT = "text"
\ No newline at end of file
+    TEXT = "text"
