diff --git a/src/core/base/types/MonorepoEntry.py b/src/core/base/types/MonorepoEntry.py
index 7947f1a..5404c4c 100644
--- a/src/core/base/types/MonorepoEntry.py
+++ b/src/core/base/types/MonorepoEntry.py
@@ -24,10 +24,14 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ChangelogEntry import ChangelogEntry
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
+
+
+
+
+
 @dataclass
 class MonorepoEntry:
     """Changelog entry for monorepo aggregation.
@@ -41,4 +45,4 @@ class MonorepoEntry:
     package_name: str
     version: str
     entries: list[ChangelogEntry] = field(default_factory=lambda: [])
-    path: str = ""
\ No newline at end of file
+    path: str = ""
