diff --git a/src/core/base/types/WCAGLevel.py b/src/core/base/types/WCAGLevel.py
index 87ca1cc..6b547d0 100644
--- a/src/core/base/types/WCAGLevel.py
+++ b/src/core/base/types/WCAGLevel.py
@@ -26,8 +26,18 @@ from enum import Enum
 
 __version__ = VERSION
 
+
+
+
+
+
+
+
+
+
+
 class WCAGLevel(Enum):
     """WCAG conformance levels."""
     A = "A"       # Minimum level
     AA = "AA"     # Mid - range level (legal requirement in many jurisdictions)
-    AAA = "AAA"   # Highest level
\ No newline at end of file
+    AAA = "AAA"   # Highest level
