diff --git a/src/logic/agents/__init__.py b/src/logic/agents/__init__.py
index 555eaea..321039d 100644
--- a/src/logic/agents/__init__.py
+++ b/src/logic/agents/__init__.py
@@ -24,4 +24,4 @@ from src.core.base.version import VERSION as VERSION
 from .swarm.OrchestratorAgent import OrchestratorAgent as Agent
 
 __all__ = ["Agent"]
-__version__ = VERSION
\ No newline at end of file
+__version__ = VERSION
