diff --git a/src/infrastructure/base/BaseManager.py b/src/infrastructure/base/BaseManager.py
index 835d396..4ad5b5d 100644
--- a/src/infrastructure/base/BaseManager.py
+++ b/src/infrastructure/base/BaseManager.py
@@ -29,4 +29,4 @@ class BaseManager(ABC):
 
     def shutdown(self) -> None:
         """Gracefully shutdown the manager."""
-        logging.debug(f"{self.__class__.__name__} shutting down.")
\ No newline at end of file
+        logging.debug(f"{self.__class__.__name__} shutting down.")
