diff --git a/src/logic/agents/intelligence/FinancialAgent.py b/src/logic/agents/intelligence/FinancialAgent.py
index 1399a48..1949bdd 100644
--- a/src/logic/agents/intelligence/FinancialAgent.py
+++ b/src/logic/agents/intelligence/FinancialAgent.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict, List, Any
+from typing import Any
 from src.core.base.BaseAgent import BaseAgent
 from src.core.base.utilities import create_main_function, as_tool
 import logging
@@ -73,4 +73,4 @@ class FinancialAgent(BaseAgent):
 
 if __name__ == "__main__":
     main = create_main_function(FinancialAgent, "Financial Agent", "File containing financial data or topic")
-    main()
\ No newline at end of file
+    main()
