diff --git a/src/infrastructure/backend/RunnerBackends.py b/src/infrastructure/backend/RunnerBackends.py
index 89755ec..713dc41 100644
--- a/src/infrastructure/backend/RunnerBackends.py
+++ b/src/infrastructure/backend/RunnerBackends.py
@@ -27,7 +27,7 @@ import os
 import subprocess
 import json
 from pathlib import Path
-from typing import Any, Optional
+from typing import Any
 
 __version__ = VERSION
 
