diff --git a/src/core/base/types/__init__.py b/src/core/base/types/__init__.py
index 95544c8..18addaf 100644
--- a/src/core/base/types/__init__.py
+++ b/src/core/base/types/__init__.py
@@ -67,4 +67,4 @@ from .TestGap import TestGap as TestGap
 from .VersioningStrategy import VersioningStrategy as VersioningStrategy
 from .WCAGLevel import WCAGLevel as WCAGLevel
 
-__version__ = VERSION
\ No newline at end of file
+__version__ = VERSION
