diff --git a/src/infrastructure/backend/__init__.py b/src/infrastructure/backend/__init__.py
index a9292a4..05b85fd 100644
--- a/src/infrastructure/backend/__init__.py
+++ b/src/infrastructure/backend/__init__.py
@@ -67,4 +67,4 @@ from .UsageRecord import UsageRecord as UsageRecord
 from .VersionNegotiator import VersionNegotiator as VersionNegotiator
 from .DiskCache import DiskCache as DiskCache
 
-__version__ = VERSION
\ No newline at end of file
+__version__ = VERSION
