diff --git a/src/core/base/utils/ExecutionScheduler.py b/src/core/base/utils/ExecutionScheduler.py
index bce41ca..0911680 100644
--- a/src/core/base/utils/ExecutionScheduler.py
+++ b/src/core/base/utils/ExecutionScheduler.py
@@ -23,11 +23,16 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.utils.ScheduledExecution import ScheduledExecution
-from typing import Optional, Dict, Any
+from typing import Any
 import time
 
 __version__ = VERSION
 
+
+
+
+
+
 class ExecutionScheduler:
     """Schedule agent executions.
 
@@ -140,4 +145,4 @@ class ExecutionScheduler:
         """
         if name in self._schedules:
             return self._schedules[name].agent_config
-        return {}
\ No newline at end of file
+        return {}
