diff --git a/src/infrastructure/dev/test_utils/TestLogEntry.py b/src/infrastructure/dev/test_utils/TestLogEntry.py
index e840bba..84e3815 100644
--- a/src/infrastructure/dev/test_utils/TestLogEntry.py
+++ b/src/infrastructure/dev/test_utils/TestLogEntry.py
@@ -23,14 +23,23 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import Any, Dict, Optional
+from typing import Any
 import time
 
 __version__ = VERSION
 
+
+
+
+
+
+
+
+
+
+
 @dataclass
 class TestLogEntry:
-    __test__ = False
     """A test log entry.
 
     Attributes:
@@ -40,9 +49,10 @@ class TestLogEntry:
         test_name: Associated test.
         extra: Extra data.
     """
+    __test__ = False
 
     level: str
     message: str
     timestamp: float = field(default_factory=time.time)
     test_name: str | None = None
-    extra: dict[str, Any] = field(default_factory=lambda: {})
\ No newline at end of file
+    extra: dict[str, Any] = field(default_factory=lambda: {})
