diff --git a/src/logic/agents/cognitive/context/utils/CodeGenerator.py b/src/logic/agents/cognitive/context/utils/CodeGenerator.py
index 51f1815..f9ef523 100644
--- a/src/logic/agents/cognitive/context/utils/CodeGenerator.py
+++ b/src/logic/agents/cognitive/context/utils/CodeGenerator.py
@@ -23,10 +23,19 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from src.logic.agents.cognitive.context.models.GeneratedCode import GeneratedCode
-from typing import Dict, List, Optional
 
 __version__ = VERSION
 
+
+
+
+
+
+
+
+
+
+
 class CodeGenerator:
     """Generates code based on context.
 
@@ -107,4 +116,4 @@ class CodeGenerator:
             code=code,
             context_used=used_contexts,
             description=prompt,
-        )
\ No newline at end of file
+        )
