diff --git a/src/core/base/deprecated.py b/src/core/base/deprecated.py
index a49f43f..a550642 100644
--- a/src/core/base/deprecated.py
+++ b/src/core/base/deprecated.py
@@ -35,6 +35,6 @@ if str(root / "src") not in sys.path:
     sys.path.append(str(root / "src"))
 
 if __name__ == '__main__':
-    # OrchestratorAgent does not have a main function. 
+    # OrchestratorAgent does not have a main function.
     # This file is deprecated.
-    pass
\ No newline at end of file
+    pass
