diff --git a/src/logic/agents/intelligence/__init__.py b/src/logic/agents/intelligence/__init__.py
index 5beeb3c..414f84f 100644
--- a/src/logic/agents/intelligence/__init__.py
+++ b/src/logic/agents/intelligence/__init__.py
@@ -23,4 +23,4 @@ from __future__ import annotations
 from src.core.base.version import VERSION as VERSION
 from .SearchAgent import SearchAgent as SearchAgent
 
-__version__ = VERSION
\ No newline at end of file
+__version__ = VERSION
