diff --git a/src/core/base/types/RefactoringPattern.py b/src/core/base/types/RefactoringPattern.py
index fe0d0fd..868306e 100644
--- a/src/core/base/types/RefactoringPattern.py
+++ b/src/core/base/types/RefactoringPattern.py
@@ -27,6 +27,11 @@ from dataclasses import dataclass
 
 __version__ = VERSION
 
+
+
+
+
+
 @dataclass
 class RefactoringPattern:
     """A code refactoring pattern."""
@@ -34,4 +39,4 @@ class RefactoringPattern:
     description: str
     pattern: str
     replacement: str
-    language: CodeLanguage = CodeLanguage.PYTHON
\ No newline at end of file
+    language: CodeLanguage = CodeLanguage.PYTHON
