diff --git a/src/observability/improvements/EffortEstimateResult.py b/src/observability/improvements/EffortEstimateResult.py
index c422fbe..e028e25 100644
--- a/src/observability/improvements/EffortEstimateResult.py
+++ b/src/observability/improvements/EffortEstimateResult.py
@@ -26,6 +26,16 @@ from dataclasses import dataclass
 
 __version__ = VERSION
 
+
+
+
+
+
+
+
+
+
+
 @dataclass
 class EffortEstimateResult:
-    hours: float
\ No newline at end of file
+    hours: float
