diff --git a/src/observability/reports/ChangelogSearcher.py b/src/observability/reports/ChangelogSearcher.py
index 199c50c..4ae3cda 100644
--- a/src/observability/reports/ChangelogSearcher.py
+++ b/src/observability/reports/ChangelogSearcher.py
@@ -22,8 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from .SearchResult import SearchResult
-from typing import List
+from src.core.base.types import SearchResult
 import re
 
 __version__ = VERSION
