diff --git a/src/core/utils/ReferenceLinkManager.py b/src/core/utils/ReferenceLinkManager.py
index 9f57e1a..94cfb2c 100644
--- a/src/core/utils/ReferenceLinkManager.py
+++ b/src/core/utils/ReferenceLinkManager.py
@@ -22,8 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from .LinkedReference import LinkedReference
-from typing import Dict, List
+from src.core.base.types import LinkedReference
 
 __version__ = VERSION
 
