diff --git a/src/core/utils/EntryReorderer.py b/src/core/utils/EntryReorderer.py
index c833901..a269604 100644
--- a/src/core/utils/EntryReorderer.py
+++ b/src/core/utils/EntryReorderer.py
@@ -22,9 +22,8 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from .ChangelogEntry import ChangelogEntry
+from src.core.base.types import ChangelogEntry
 from .GroupingStrategy import GroupingStrategy
-from typing import Dict, List
 
 __version__ = VERSION
 
