diff --git a/src/observability/reports/ChangelogLocalizer.py b/src/observability/reports/ChangelogLocalizer.py
index 0a18a14..da614bd 100644
--- a/src/observability/reports/ChangelogLocalizer.py
+++ b/src/observability/reports/ChangelogLocalizer.py
@@ -22,9 +22,8 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from .LocalizationLanguage import LocalizationLanguage
-from .LocalizedEntry import LocalizedEntry
-from typing import List
+from src.core.base.types import LocalizationLanguage
+from src.core.base.types import LocalizedEntry
 
 __version__ = VERSION
 
