diff --git a/src/observability/improvements/DocGenerator.py b/src/observability/improvements/DocGenerator.py
index 4a44c2a..7cb8932 100644
--- a/src/observability/improvements/DocGenerator.py
+++ b/src/observability/improvements/DocGenerator.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .Improvement import Improvement
-from typing import Any, Dict, cast
+from typing import Any, cast
 
 __version__ = VERSION
 
