diff --git a/src/infrastructure/dev/test_utils/EnvironmentIsolator.py b/src/infrastructure/dev/test_utils/EnvironmentIsolator.py
index 9a5d137..d2cb56c 100644
--- a/src/infrastructure/dev/test_utils/EnvironmentIsolator.py
+++ b/src/infrastructure/dev/test_utils/EnvironmentIsolator.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Any, Dict
+from typing import Any
 import os
 
 __version__ = VERSION
