diff --git a/src/observability/improvements/ArchiveManager.py b/src/observability/improvements/ArchiveManager.py
index 3a418a8..802ddc9 100644
--- a/src/observability/improvements/ArchiveManager.py
+++ b/src/observability/improvements/ArchiveManager.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .Improvement import Improvement
-from typing import List
 
 __version__ = VERSION
 
