diff --git a/src/observability/reports/ReportTemplate.py b/src/observability/reports/ReportTemplate.py
index 0cf4e99..d58ac78 100644
--- a/src/observability/reports/ReportTemplate.py
+++ b/src/observability/reports/ReportTemplate.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
