diff --git a/src/infrastructure/backend/UsageQuotaManager.py b/src/infrastructure/backend/UsageQuotaManager.py
index f46f376..6d27541 100644
--- a/src/infrastructure/backend/UsageQuotaManager.py
+++ b/src/infrastructure/backend/UsageQuotaManager.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .UsageQuota import UsageQuota
-from typing import Any, Dict, Tuple
+from typing import Any
 import threading
 import time
 
