diff --git a/src/observability/improvements/BranchComparer.py b/src/observability/improvements/BranchComparer.py
index d006a0f..bbd685c 100644
--- a/src/observability/improvements/BranchComparer.py
+++ b/src/observability/improvements/BranchComparer.py
@@ -28,7 +28,7 @@ from .Improvement import Improvement
 from .ImprovementDiff import ImprovementDiff
 from .ImprovementDiffType import ImprovementDiffType
 from pathlib import Path
-from typing import Any, Dict, List, Optional, Tuple
+from typing import Any
 import hashlib
 import logging
 import re
