diff --git a/src/logic/agents/cognitive/context/utils/ContextInheritance.py b/src/logic/agents/cognitive/context/utils/ContextInheritance.py
index fc5c7f8..2663918 100644
--- a/src/logic/agents/cognitive/context/utils/ContextInheritance.py
+++ b/src/logic/agents/cognitive/context/utils/ContextInheritance.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from src.logic.agents.cognitive.context.models.InheritanceMode import InheritanceMode
 from src.logic.agents.cognitive.context.models.InheritedContext import InheritedContext
-from typing import Dict, List, Optional
 import re
 
 __version__ = VERSION
