diff --git a/src/observability/improvements/ScheduledImprovement.py b/src/observability/improvements/ScheduledImprovement.py
index 908fc26..c578147 100644
--- a/src/observability/improvements/ScheduledImprovement.py
+++ b/src/observability/improvements/ScheduledImprovement.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ScheduleStatus import ScheduleStatus
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
