diff --git a/src/infrastructure/dev/test_utils/RetryHelper.py b/src/infrastructure/dev/test_utils/RetryHelper.py
index 6137ac7..0862c82 100644
--- a/src/infrastructure/dev/test_utils/RetryHelper.py
+++ b/src/infrastructure/dev/test_utils/RetryHelper.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Optional, TypeVar
+from typing import TypeVar
 from collections.abc import Callable
 import threading
 
