diff --git a/src/infrastructure/dev/test_utils/Benchmarker.py b/src/infrastructure/dev/test_utils/Benchmarker.py
index 84134cd..5a92c09 100644
--- a/src/infrastructure/dev/test_utils/Benchmarker.py
+++ b/src/infrastructure/dev/test_utils/Benchmarker.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .TestTimer import TestTimer
-from typing import Dict, List
 from collections.abc import Callable
 
 __version__ = VERSION
