diff --git a/src/infrastructure/sandbox/SandboxRuntime.py b/src/infrastructure/sandbox/SandboxRuntime.py
index 4af28f2..3453bdb 100644
--- a/src/infrastructure/sandbox/SandboxRuntime.py
+++ b/src/infrastructure/sandbox/SandboxRuntime.py
@@ -1,6 +1,6 @@
 
 import logging
-from typing import Any, Dict
+from typing import Any
 from src.infrastructure.sandbox.core.SandboxCore import SandboxCore
 
 class SandboxRuntime:
