diff --git a/src/observability/errors/ExternalReportingClient.py b/src/observability/errors/ExternalReportingClient.py
index 475bb1a..160199e 100644
--- a/src/observability/errors/ExternalReportingClient.py
+++ b/src/observability/errors/ExternalReportingClient.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ErrorEntry import ErrorEntry
 from .ExternalReporter import ExternalReporter
-from typing import Any, Dict, List
+from typing import Any
 import logging
 
 __version__ = VERSION
