diff --git a/src/observability/improvements/ResourceAllocation.py b/src/observability/improvements/ResourceAllocation.py
index fb1e2fb..e3dc4bb 100644
--- a/src/observability/improvements/ResourceAllocation.py
+++ b/src/observability/improvements/ResourceAllocation.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
