diff --git a/src/observability/errors/ErrorCluster.py b/src/observability/errors/ErrorCluster.py
index dd48f0f..c1b75a6 100644
--- a/src/observability/errors/ErrorCluster.py
+++ b/src/observability/errors/ErrorCluster.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
