diff --git a/src/infrastructure/dev/test_utils/RecordedInteraction.py b/src/infrastructure/dev/test_utils/RecordedInteraction.py
index 694b345..7814b26 100644
--- a/src/infrastructure/dev/test_utils/RecordedInteraction.py
+++ b/src/infrastructure/dev/test_utils/RecordedInteraction.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import Any, Dict, Tuple
+from typing import Any
 import time
 
 __version__ = VERSION
