diff --git a/src/core/base/types/ModernizationSuggestion.py b/src/core/base/types/ModernizationSuggestion.py
index bf3d9b4..de3748c 100644
--- a/src/core/base/types/ModernizationSuggestion.py
+++ b/src/core/base/types/ModernizationSuggestion.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass
-from typing import Optional
 
 __version__ = VERSION
 
