diff --git a/src/infrastructure/dev/test_utils/ParameterizedTestGenerator.py b/src/infrastructure/dev/test_utils/ParameterizedTestGenerator.py
index dda3acb..8b31a9c 100644
--- a/src/infrastructure/dev/test_utils/ParameterizedTestGenerator.py
+++ b/src/infrastructure/dev/test_utils/ParameterizedTestGenerator.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .ParameterizedTestCase import ParameterizedTestCase
-from typing import Any, Dict, List, Optional
+from typing import Any
 from collections.abc import Callable
 
 __version__ = VERSION
