diff --git a/src/observability/reports/AuditEntry.py b/src/observability/reports/AuditEntry.py
index 898ca1a..0eeeb74 100644
--- a/src/observability/reports/AuditEntry.py
+++ b/src/observability/reports/AuditEntry.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .AuditAction import AuditAction
 from dataclasses import dataclass, field
-from typing import Any, Dict
+from typing import Any
 
 __version__ = VERSION
 
