diff --git a/src/observability/reports/ReportFilter.py b/src/observability/reports/ReportFilter.py
index 9f7457e..43a73d1 100644
--- a/src/observability/reports/ReportFilter.py
+++ b/src/observability/reports/ReportFilter.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .CodeIssue import CodeIssue
 from .FilterCriteria import FilterCriteria
-from typing import List, Optional
 
 __version__ = VERSION
 
