diff --git a/src/core/base/utils/RateLimiter.py b/src/core/base/utils/RateLimiter.py
index 3c57bb4..997325d 100644
--- a/src/core/base/utils/RateLimiter.py
+++ b/src/core/base/utils/RateLimiter.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from src.core.base.models import RateLimitConfig
-from typing import List, Optional, Dict, Any
+from typing import Any
 import threading
 import time
 
