diff --git a/src/infrastructure/dev/test_utils/TestLogEntry.py b/src/infrastructure/dev/test_utils/TestLogEntry.py
index e840bba..5a060e1 100644
--- a/src/infrastructure/dev/test_utils/TestLogEntry.py
+++ b/src/infrastructure/dev/test_utils/TestLogEntry.py
@@ -30,7 +30,6 @@ __version__ = VERSION
 
 @dataclass
 class TestLogEntry:
-    __test__ = False
     """A test log entry.
 
     Attributes:
@@ -40,6 +39,7 @@ class TestLogEntry:
         test_name: Associated test.
         extra: Extra data.
     """
+    __test__ = False
 
     level: str
     message: str
