diff --git a/src/logic/agents/cognitive/context/models/ContextVersion.py b/src/logic/agents/cognitive/context/models/ContextVersion.py
index a719244..02e8294 100644
--- a/src/logic/agents/cognitive/context/models/ContextVersion.py
+++ b/src/logic/agents/cognitive/context/models/ContextVersion.py
@@ -22,7 +22,6 @@
 
 from __future__ import annotations
 from dataclasses import dataclass, field
-from typing import List
 
 @dataclass
 class ContextVersion:
