diff --git a/src/core/base/utils/ConditionalExecutor.py b/src/core/base/utils/ConditionalExecutor.py
index 21f44f6..caf805b 100644
--- a/src/core/base/utils/ConditionalExecutor.py
+++ b/src/core/base/utils/ConditionalExecutor.py
@@ -22,9 +22,9 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from .ExecutionCondition import ExecutionCondition
+from src.core.base.models.base_models import ExecutionCondition
 from pathlib import Path
-from typing import List, Dict, Any
+from typing import Any
 from collections.abc import Callable
 
 __version__ = VERSION
