diff --git a/src/observability/errors/BranchComparer.py b/src/observability/errors/BranchComparer.py
index a2731c3..dec31b2 100644
--- a/src/observability/errors/BranchComparer.py
+++ b/src/observability/errors/BranchComparer.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .BranchComparison import BranchComparison
-from typing import Dict, List, Set
 
 __version__ = VERSION
 
