diff --git a/src/infrastructure/dev/test_utils/MockAIBackend.py b/src/infrastructure/dev/test_utils/MockAIBackend.py
index d0140ba..0da9e23 100644
--- a/src/infrastructure/dev/test_utils/MockAIBackend.py
+++ b/src/infrastructure/dev/test_utils/MockAIBackend.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .MockResponse import MockResponse
 from .MockResponseType import MockResponseType
-from typing import Dict, List, Tuple, Optional
 import logging
 import re
 import time
