diff --git a/src/core/base/utils/AgentPriorityQueue.py b/src/core/base/utils/AgentPriorityQueue.py
index 6e50dc1..7daedd7 100644
--- a/src/core/base/utils/AgentPriorityQueue.py
+++ b/src/core/base/utils/AgentPriorityQueue.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import List, Optional, Dict, Any
+from typing import Any
 import logging
 
 __version__ = VERSION
