diff --git a/src/observability/reports/AggregatedReport.py b/src/observability/reports/AggregatedReport.py
index e6cc5f4..93e02d0 100644
--- a/src/observability/reports/AggregatedReport.py
+++ b/src/observability/reports/AggregatedReport.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .CodeIssue import CodeIssue
 from dataclasses import dataclass, field
-from typing import Any, Dict, List
+from typing import Any
 import time
 
 __version__ = VERSION
