diff --git a/src/infrastructure/dev/test_utils/LogCapturer.py b/src/infrastructure/dev/test_utils/LogCapturer.py
index 7957091..dfe6321 100644
--- a/src/infrastructure/dev/test_utils/LogCapturer.py
+++ b/src/infrastructure/dev/test_utils/LogCapturer.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Any, List, Optional
+from typing import Any
 import logging
 
 __version__ = VERSION
