diff --git a/src/infrastructure/dev/test_utils/ParallelTestResult.py b/src/infrastructure/dev/test_utils/ParallelTestResult.py
index ce5fca9..c82cc94 100644
--- a/src/infrastructure/dev/test_utils/ParallelTestResult.py
+++ b/src/infrastructure/dev/test_utils/ParallelTestResult.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass
-from typing import Optional
 
 __version__ = VERSION
 
