diff --git a/src/core/base/managers/ConversationManagers.py b/src/core/base/managers/ConversationManagers.py
index dd3c3f9..3789df0 100644
--- a/src/core/base/managers/ConversationManagers.py
+++ b/src/core/base/managers/ConversationManagers.py
@@ -20,7 +20,6 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import List
 from src.core.base.models import MessageRole, ConversationMessage
 
 __version__ = VERSION
