diff --git a/src/logic/agents/cognitive/context/models/InheritedContext.py b/src/logic/agents/cognitive/context/models/InheritedContext.py
index 6626062..1d81610 100644
--- a/src/logic/agents/cognitive/context/models/InheritedContext.py
+++ b/src/logic/agents/cognitive/context/models/InheritedContext.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from src.logic.agents.cognitive.context.models.InheritanceMode import InheritanceMode
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
