diff --git a/src/observability/reports/ReportComparator.py b/src/observability/reports/ReportComparator.py
index ee39260..ab09a1b 100644
--- a/src/observability/reports/ReportComparator.py
+++ b/src/observability/reports/ReportComparator.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ReportComparison import ReportComparison
 from pathlib import Path
-from typing import List
 
 __version__ = VERSION
 
