diff --git a/src/infrastructure/dev/test_utils/CoverageTracker.py b/src/infrastructure/dev/test_utils/CoverageTracker.py
index 429d38c..f7020eb 100644
--- a/src/infrastructure/dev/test_utils/CoverageTracker.py
+++ b/src/infrastructure/dev/test_utils/CoverageTracker.py
@@ -22,7 +22,6 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict, Set
 
 __version__ = VERSION
 
