diff --git a/src/logic/agents/cognitive/context/utils/ContextVisualizer.py b/src/logic/agents/cognitive/context/utils/ContextVisualizer.py
index c859000..ce5e330 100644
--- a/src/logic/agents/cognitive/context/utils/ContextVisualizer.py
+++ b/src/logic/agents/cognitive/context/utils/ContextVisualizer.py
@@ -25,7 +25,7 @@ from src.core.base.version import VERSION
 from src.logic.agents.cognitive.context.models.VisualizationData import VisualizationData
 from src.logic.agents.cognitive.context.models.VisualizationType import VisualizationType
 from pathlib import Path
-from typing import Any, Dict, List, Optional, Tuple
+from typing import Any
 import json
 
 __version__ = VERSION
