diff --git a/src/infrastructure/dev/test_utils/FlakinessReport.py b/src/infrastructure/dev/test_utils/FlakinessReport.py
index e3ef62a..1758035 100644
--- a/src/infrastructure/dev/test_utils/FlakinessReport.py
+++ b/src/infrastructure/dev/test_utils/FlakinessReport.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import List
 
 __version__ = VERSION
 
