diff --git a/src/core/base/managers/AuthManagers.py b/src/core/base/managers/AuthManagers.py
index e901a1e..70ad17d 100644
--- a/src/core/base/managers/AuthManagers.py
+++ b/src/core/base/managers/AuthManagers.py
@@ -22,7 +22,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 import logging
 from dataclasses import dataclass, field
-from typing import Dict, Optional
 from src.core.base.models import AuthConfig, AuthMethod, _empty_dict_str_str
 
 __version__ = VERSION
