diff --git a/src/observability/errors/NotificationManager.py b/src/observability/errors/NotificationManager.py
index f4d052f..d10576d 100644
--- a/src/observability/errors/NotificationManager.py
+++ b/src/observability/errors/NotificationManager.py
@@ -25,7 +25,6 @@ from src.core.base.version import VERSION
 from .ErrorEntry import ErrorEntry
 from .NotificationChannel import NotificationChannel
 from .NotificationConfig import NotificationConfig
-from typing import List
 import logging
 
 __version__ = VERSION
