diff --git a/src/core/base/types/TemplateManager.py b/src/core/base/types/TemplateManager.py
index fb7c4c2..3bb1aea 100644
--- a/src/core/base/types/TemplateManager.py
+++ b/src/core/base/types/TemplateManager.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .EntryTemplate import EntryTemplate
-from typing import Dict, List
 import re
 
 __version__ = VERSION
