diff --git a/src/core/base/utils/FileLockManager.py b/src/core/base/utils/FileLockManager.py
index b3cb42a..e16142c 100644
--- a/src/core/base/utils/FileLockManager.py
+++ b/src/core/base/utils/FileLockManager.py
@@ -25,7 +25,6 @@ from src.core.base.version import VERSION
 from src.core.base.utils.FileLock import FileLock
 from src.core.base.models import LockType
 from pathlib import Path
-from typing import Optional, Dict
 import logging
 import os
 import threading
