diff --git a/src/infrastructure/backend/ABTestVariant.py b/src/infrastructure/backend/ABTestVariant.py
index 41ffc38..0288fe0 100644
--- a/src/infrastructure/backend/ABTestVariant.py
+++ b/src/infrastructure/backend/ABTestVariant.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from dataclasses import dataclass, field
-from typing import Dict
 
 __version__ = VERSION
 
