diff --git a/src/infrastructure/dev/test_utils/DependencyContainer.py b/src/infrastructure/dev/test_utils/DependencyContainer.py
index 395535a..b4b9ff2 100644
--- a/src/infrastructure/dev/test_utils/DependencyContainer.py
+++ b/src/infrastructure/dev/test_utils/DependencyContainer.py
@@ -22,7 +22,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Any, Dict, Tuple, TypeVar
+from typing import Any, TypeVar
 from collections.abc import Callable
 import logging
 import inspect
