diff --git a/src/observability/reports/FilterCriteria.py b/src/observability/reports/FilterCriteria.py
index f931e8a..0b57507 100644
--- a/src/observability/reports/FilterCriteria.py
+++ b/src/observability/reports/FilterCriteria.py
@@ -26,7 +26,6 @@ from .IssueCategory import IssueCategory
 from .SeverityLevel import SeverityLevel
 from dataclasses import dataclass
 from datetime import datetime
-from typing import List, Optional
 
 __version__ = VERSION
 
