diff --git a/src/observability/stats/observability_core.py b/src/observability/stats/observability_core.py
index 5cb9c33..54faff4 100644
--- a/src/observability/stats/observability_core.py
+++ b/src/observability/stats/observability_core.py
@@ -10,7 +10,7 @@ import zlib
 from enum import Enum
 from datetime import datetime, timedelta
 from dataclasses import dataclass, field
-from typing import Dict, List, Any, Optional, Tuple, Union
+from typing import Any
 
 try:
     import matplotlib.pyplot as plt
