diff --git a/src/core/base/types/LocalizedEntry.py b/src/core/base/types/LocalizedEntry.py
index f523a87..469ce2d 100644
--- a/src/core/base/types/LocalizedEntry.py
+++ b/src/core/base/types/LocalizedEntry.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .LocalizationLanguage import LocalizationLanguage
 from dataclasses import dataclass, field
-from typing import Dict
 
 __version__ = VERSION
 
