diff --git a/src/infrastructure/backend/ABTester.py b/src/infrastructure/backend/ABTester.py
index 6e5f649..e912fb7 100644
--- a/src/infrastructure/backend/ABTester.py
+++ b/src/infrastructure/backend/ABTester.py
@@ -23,7 +23,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .ABTestVariant import ABTestVariant
-from typing import Any, Dict, Optional, Tuple
+from typing import Any
 import threading
 
 __version__ = VERSION
