diff --git a/src/infrastructure/backend/LoadBalancer.py b/src/infrastructure/backend/LoadBalancer.py
index 7009ce8..c47f5bb 100644
--- a/src/infrastructure/backend/LoadBalancer.py
+++ b/src/infrastructure/backend/LoadBalancer.py
@@ -25,7 +25,7 @@ from src.core.base.version import VERSION
 from .SystemConfig import SystemConfig
 from .ProviderType import ProviderType
 from .LoadBalanceStrategy import LoadBalanceStrategy
-from typing import Any, Dict, List, Optional
+from typing import Any
 import logging
 import threading
 
