diff --git a/src/core/utils/ReferenceLinkManager.py b/src/core/utils/ReferenceLinkManager.py
index 9f57e1a..8bb5817 100644
--- a/src/core/utils/ReferenceLinkManager.py
+++ b/src/core/utils/ReferenceLinkManager.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .LinkedReference import LinkedReference
-from typing import Dict, List
 
 __version__ = VERSION
 
