diff --git a/src/logic/agents/cognitive/context/engines/GlobalContextCore.py b/src/logic/agents/cognitive/context/engines/GlobalContextCore.py
index d20f540..42290eb 100644
--- a/src/logic/agents/cognitive/context/engines/GlobalContextCore.py
+++ b/src/logic/agents/cognitive/context/engines/GlobalContextCore.py
@@ -19,7 +19,7 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict, List, Any
+from typing import Any
 from datetime import datetime
 
 __version__ = VERSION
