diff --git a/src/infrastructure/backend/SystemConfig.py b/src/infrastructure/backend/SystemConfig.py
index 83c5cc1..9ec8c5d 100644
--- a/src/infrastructure/backend/SystemConfig.py
+++ b/src/infrastructure/backend/SystemConfig.py
@@ -24,7 +24,6 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .ProviderType import ProviderType
 from dataclasses import dataclass
-from typing import Optional
 
 __version__ = VERSION
 
