diff --git a/src/infrastructure/backend/RequestContext.py b/src/infrastructure/backend/RequestContext.py
index ea06637..8a32d53 100644
--- a/src/infrastructure/backend/RequestContext.py
+++ b/src/infrastructure/backend/RequestContext.py
@@ -24,7 +24,7 @@ from __future__ import annotations
 from src.core.base.version import VERSION
 from .RequestPriority import RequestPriority
 from dataclasses import dataclass, field
-from typing import Any, Dict, Optional
+from typing import Any
 import time
 import uuid
 
