diff --git a/src/core/base/utils/AgentGitHandler.py b/src/core/base/utils/AgentGitHandler.py
index 0fe58f4..c4f4b93 100644
--- a/src/core/base/utils/AgentGitHandler.py
+++ b/src/core/base/utils/AgentGitHandler.py
@@ -23,7 +23,7 @@ from src.core.base.version import VERSION
 import subprocess
 import logging
 from pathlib import Path
-from typing import List, Optional, Any, Dict
+from typing import Any
 
 __version__ = VERSION
 
