diff --git a/src/observability/reports/AnnotationManager.py b/src/observability/reports/AnnotationManager.py
index 05a445e..bbcff16 100644
--- a/src/observability/reports/AnnotationManager.py
+++ b/src/observability/reports/AnnotationManager.py
@@ -23,7 +23,6 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .ReportAnnotation import ReportAnnotation
-from typing import Dict, List, Optional
 import logging
 
 __version__ = VERSION
