diff --git a/src/logic/agents/system/CloudProviderAgent.py b/src/logic/agents/system/CloudProviderAgent.py
index 93e2990..decee29 100644
--- a/src/logic/agents/system/CloudProviderAgent.py
+++ b/src/logic/agents/system/CloudProviderAgent.py
@@ -19,7 +19,6 @@
 
 from __future__ import annotations
 from src.core.base.version import VERSION
-from typing import Dict
 from src.core.base.BaseAgent import BaseAgent
 
 __version__ = VERSION
