diff --git a/src/logic/strategies/ChainOfThoughtStrategy.py b/src/logic/strategies/ChainOfThoughtStrategy.py
index 3bcefc7..e646ce1 100644
--- a/src/logic/strategies/ChainOfThoughtStrategy.py
+++ b/src/logic/strategies/ChainOfThoughtStrategy.py
@@ -5,7 +5,7 @@
 from __future__ import annotations
 from src.core.base.version import VERSION
 from .AgentStrategy import AgentStrategy
-from typing import Dict, List, Optional, TYPE_CHECKING
+from typing import TYPE_CHECKING
 import logging
 
 if TYPE_CHECKING:
